Source/Dafny/Compilers/Compiler.cs

```diff
@@ -1194,6 +1196,8 @@ public enum ResolvedUnaryOp { BoolNot, BitwiseNot, Cardinality }
             var w = DeclareDatatype(dt, wr);
             if (w != null) {
               CompileClassMembers(program, dt, w);
+            } else if (this is not CppCompiler) {
```

Comment:
The alternative is living with a `\n` when there's a tuple. I personally don't like either.

Reply:
Could this be an attribute of each of the compilers that the main code checks to decide whether to add the `\n`?

Reply:
This code is a little strange, we always add a `\n` in L1161 for every TopLevelDecl that's compiled. Sometimes we only find out later that we don't have to do anything. This leaves unsightly whitespace in the generated sources. Looking it is no fun.

The concrete problem here is a hack in the cpp compiler. They include the ClassMembers on the first pass, so they can't have the second one. To avoid that, they act like they didn't do anything.

Reply:
Sounds like we can still define an attribute, even if it's a weird name like `CompilesClassMembersOnTheFirstPass` or whatever. :)

Referring to concrete subclasses here is just a recipe for weird hard-to-debug edge cases as we add more compilers.

Reply:
Now, it's this one field we happen to overwrite in this one compiler.